thrust/thrust/iterator/iterator_traits.h

```diff
@@ -194,6 +195,13 @@ template <typename Iterator>
 using iterator_system_t = typename iterator_system<Iterator>::type;
 
 // specialize the respective cuda iterators
+// we need to be able to pass a system to `cuda::counting_iterator`.
+template <class System>
+using __cccl_translate_system = ::cuda::std::conditional_t<
+  ::cuda::std::is_same_v<System, ::cuda::__cccl_any_system_tag> || ::cuda::std::is_same_v<System, use_default>,
+  any_system_tag,
+  System>;
+
```

Comment:
I don't understand why we need this. Counting iterators work on any system. Also, a `counting_iterator` should never be instantiated with `use_default`, because that tag is part of the iterator facade/adaptors and should already be resolved there. What problem are we solving here?

Reply:
After some offline discussion, we identified a problem with how iterator categories are deduced: #5860. We should fix this before continuing with this PR.